crates/cargo-bsp/src/cargo_communication/request_actor.rs

```diff
@@ -898,4 +898,468 @@ pub mod compile_request_tests {
+                  "parents": "random_task_id"
+                },
+                "total": 0,
+                "unit": "tests"
+              }
+            }
+            "###);
+            no_more_msg(receiver_from_actor);
+        }
+
+        #[test]
+        fn test_finished_failed() {
+            let (mut req_actor, receiver_from_actor, _cancel_sender) =
+                default_req_actor::<Test>(MockCargoHandler::new(), default_test_params());
+
+            let test_started = Started(TestName {
+                name: TEST_NAME.into(),
+            });
+            req_actor.handle_event(CargoEvent(CargoStdout(TextLine(
+                to_string(&TestType::Test(test_started)).unwrap(),
+            ))));
+            let _ = receiver_from_actor.recv().unwrap(); // test started message
+
+            let test_finished_ok = TestEvent::Failed(TestResultEnum {
+                name: TEST_NAME.to_string(),
+                stdout: None,
+            });
+            req_actor.handle_event(CargoEvent(CargoStdout(TextLine(
+                to_string(&TestType::Test(test_finished_ok)).unwrap(),
+            ))));
+
+            assert_json_snapshot!(receiver_from_actor.recv().unwrap(), {
+                ".params.taskId.id" => RANDOM_TASK_ID,
+                ".params.taskId.parents" => format!("{RANDOM_TASK_ID}"),
+                ".params.eventTime" => TIMESTAMP,
+            } ,@r###"
+            {
+              "method": "build/taskFinish",
+              "params": {
+                "data": {
+                  "displayName": "test_name",
+                  "status": 2
+                },
+                "dataKind": "test-finish",
+                "eventTime": "timestamp",
+                "status": 1,
+                "taskId": {
+                  "id": "random_task_id",
+                  "parents": "random_task_id"
+                }
+              }
+            }
+            "###);
+            let _ = receiver_from_actor.recv().unwrap(); // test progress message
+            no_more_msg(receiver_from_actor);
+        }
+
+        #[test]
+        fn test_finished_ignored() {
+            let (mut req_actor, receiver_from_actor, _cancel_sender) =
+                default_req_actor::<Test>(MockCargoHandler::new(), default_test_params());
+
+            let test_started = Started(TestName {
+                name: TEST_NAME.into(),
+            });
+            req_actor.handle_event(CargoEvent(CargoStdout(TextLine(
+                to_string(&TestType::Test(test_started)).unwrap(),
+            ))));
+            let _ = receiver_from_actor.recv().unwrap(); // test started message
+
+            let test_finished_ok = TestEvent::Ignored(TestResultEnum {
+                name: TEST_NAME.to_string(),
+                stdout: None,
+            });
+            req_actor.handle_event(CargoEvent(CargoStdout(TextLine(
+                to_string(&TestType::Test(test_finished_ok)).unwrap(),
+            ))));
```

Comment:
Może by wydzielić to do jakiejś osobnej funkcji, skoro powtarza się praktycznie to samo w czterech testach? Można by też pomyśleć czy całych tych testów nie da się uogólnić -> różnią się tylko wiadomością przekazywaną do handle_event oraz polem status w oczekiwanej odpowiedzi (a przy failed i timeout to nawet to się nie różni)